### DISCUSSION TITLE
How to create triggers in mysql db using prisma migrate

### DISCUSSION THREAD

**Discussion by karan-c3** [prisma/prisma#9611] [Q&A] [answered]
Hi,
I am trying to create trigger using migrate, i have following code in migration files
```sql
-- AlterTable
DELIMITER $$
CREATE TRIGGER `Order_BEFORE_INSERT` BEFORE INSERT ON `Order` FOR EACH ROW BEGIN
declare  v_doc_no varchar(100);
declare v_prefix varchar(10);
SELECT 
    CASE
        WHEN new.docType = 'Sales Invoice' THEN 'INV-'
        WHEN new.docType = 'Fulfillment' THEN ' SHIP-'
        WHEN new.docType = 'Sales Order' THEN 'SO'
        WHEN new.docType = 'Sales Quote' THEN 'SQ-'
        WHEN new.docType = 'Sales Receipt' THEN 'PREPAY-'
        WHEN new.docType = 'Sales Return' THEN 'RET-' 
        WHEN new.docType = 'Shipper' THEN 'SHIP-'
        WHEN new.docType = 'Web Order' THEN ' WO-'
        WHEN new.docType = 'Blanket Order' THEN 'BK-'
        WHEN new.docType = 'Proforma Invoice' THEN 'PI'
        ELSE ''
    END
INTO v_prefix;

SELECT  CONCAT(v_prefix, MAX(id) + 1) INTO v_doc_no FROM Order;
set New.docNo = v_doc_no; 
END$$
DELIMITER ;
```
But after applying migration I get following error, I've also tried without delimiters still no luck 
```bash
Error: P3018

A migration failed to apply. New migrations cannot be applied before the error is recovered from. Read more about how to resolve migration issues in a production database: https://pris.ly/d/migrate-resolve

Migration name: 20211005120033_adding_trigger_for_doc_no

Database error code: 1064

Database error:
You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'DELIMITER $$
CREATE TRIGGER `Order_BEFORE_INSERT` BEFORE INSERT ON `Order` FOR E' at line 2

Please check the query number 1 from the migration file.
```

**Comment by karan-c3** [comment, marked as answer]
got it, need to put Order in Backticks ``` `Order` ``` since the order is a reserved keyword in MySQL

**Reply by Veercodeprog** [reply]
i am getting error for everyline in the trigger code. i am using mysql database

model popular_now {
  popular_now_id             Int      @id @default(autoincrement())
  name           String
  is_product     Boolean
  display_text   String
  hyperlink      String?
  is_active      Boolean
  reference_id   Int?
 
 @@db.execute(
    """
    CREATE TRIGGER category_added_trigger AFTER INSERT ON Category
    FOR EACH ROW
    BEGIN
      INSERT INTO PopularNow (name, is_product, display_text, hyperlink, is_active,reference_id)
      VALUES (NEW.name, false, NEW.name, NULL, false, New.category_id);
    END;
    """
  )

  @@db.execute(
    """
    CREATE TRIGGER product_added_trigger AFTER INSERT ON Product
    FOR EACH ROW
    BEGIN
      INSERT INTO PopularNow (name, is_product, display_text, hyperlink, is_active, reference_id)
      VALUES (NEW.name, true, NEW.name, NEW.website_url, false, NEW.product_id);

    """
  )
}

**Reply by Veercodeprog** [reply]
Error: Prisma schema validation - (get-config wasm)
Error code: P1012
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:47
   | 
46 |  
47 |  @@db.execute(
48 |     """
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:48
   | 
47 |  @@db.execute(
48 |     """
49 |     CREATE TRIGGER category_added_trigger AFTER INSERT ON Category
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:49
   | 
48 |     """
49 |     CREATE TRIGGER category_added_trigger AFTER INSERT ON Category
50 |     FOR EACH ROW
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:50
   | 
49 |     CREATE TRIGGER category_added_trigger AFTER INSERT ON Category
50 |     FOR EACH ROW
51 |     BEGIN
   | 
error: Error validating model "popular_now": This field declaration is invalid. It is either missing a name or a type.
  -->  schema.prisma:51
   | 
50 |     FOR EACH ROW
51 |     BEGIN
52 |       INSERT INTO PopularNow (name, is_product, display_text, hyperlink, is_active,reference_id)
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:52
   | 
51 |     BEGIN
52 |       INSERT INTO PopularNow (name, is_product, display_text, hyperlink, is_active,reference_id)
53 |       VALUES (NEW.name, false, NEW.name, NULL, false, New.category_id);
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:53
   | 
52 |       INSERT INTO PopularNow (name, is_product, display_text, hyperlink, is_active,reference_id)
53 |       VALUES (NEW.name, false, NEW.name, NULL, false, New.category_id);
54 |     END;
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:54
   | 
53 |       VALUES (NEW.name, false, NEW.name, NULL, false, New.category_id);
54 |     END;
55 |     """
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:55
   | 
54 |     END;
55 |     """
56 |   )
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:56
   | 
55 |     """
56 |   )
57 | 
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:58
   | 
57 | 
58 |   @@db.execute(
59 |     """
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:59
   | 
58 |   @@db.execute(
59 |     """
60 |     CREATE TRIGGER product_added_trigger AFTER INSERT ON Product
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:60
   | 
59 |     """
60 |     CREATE TRIGGER product_added_trigger AFTER INSERT ON Product
61 |     FOR EACH ROW
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:61
   | 
60 |     CREATE TRIGGER product_added_trigger AFTER INSERT ON Product
61 |     FOR EACH ROW
62 |     BEGIN
   | 
error: Error validating model "popular_now": This field declaration is invalid. It is either missing a name or a type.
  -->  schema.prisma:62
   | 
61 |     FOR EACH ROW
62 |     BEGIN
63 |       INSERT INTO PopularNow (name, is_product, display_text, hyperlink, is_active, reference_id)
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:63
   | 
62 |     BEGIN
63 |       INSERT INTO PopularNow (name, is_product, display_text, hyperlink, is_active, reference_id)
64 |       VALUES (NEW.name, true, NEW.name, NEW.website_url, false, NEW.product_id);
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:64
   | 
63 |       INSERT INTO PopularNow (name, is_product, display_text, hyperlink, is_active, reference_id)
64 |       VALUES (NEW.name, true, NEW.name, NEW.website_url, false, NEW.product_id);
65 |     END;
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:65
   | 
64 |       VALUES (NEW.name, true, NEW.name, NEW.website_url, false, NEW.product_id);
65 |     END;
66 |     """
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:66
   | 
65 |     END;
66 |     """
67 |   )
   | 
error: Error validating: This line is not a valid field or attribute definition.
  -->  schema.prisma:67
   | 
66 |     """
67 |   )
68 | }
   | 

Validation Error Count: 20
[Context: getConfig]

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
